capstone/capapi/permissions.py

```diff
@@ -18,9 +19,41 @@ def has_permission(self, request, view):
             # boot if user tries anything but `GET`, `OPTIONS`, or `HEAD`
             return False
 
-        # give away metadata to everyone!
-        if not request.query_params.get('type') == 'download':
+        # check whether user wants to get full case text
+        context = view.get_renderer_context()
+        full_case = context.get('request').query_params.get('full_case', 'false').lower()
+        if full_case != 'true':
+            # serve metadata to everyone!
             return True
+        kwargs = context.get('kwargs')
+        # get blacklisted case count
+        cite = context.get('request').query_params.get('cite', None)
+
+        if view.lookup_field in kwargs and '-' in slugify(kwargs[view.lookup_field]):
+            # assume this is a lookup using a citation if hyphen is present
+            cases = view.queryset.filter(citation__normalized_cite=slugify(kwargs[view.lookup_field]),
+                                         jurisdiction__whitelisted=False)
+        elif cite:
+            cases = view.queryset.filter(citation__normalized_cite=slugify(cite),
+                                         jurisdiction__whitelisted=False)
         else:
-            # make sure user has auth token before proceeding if they want to download
-            return request.auth
+            cases = view.queryset.filter(**kwargs, jurisdiction__whitelisted=False)
+
+        blacklisted_count = cases.count()
+
+        # If the request contains only whitelisted jurisdictions
+        # serve up whether the user is signed in or not
+        if not blacklisted_count:
+            return True
+
+        # from here forth one cannot be anonymous
+        if request.user.is_anonymous:
+            return False
+
+        is_allowed = request.user.case_allowance_remaining >= blacklisted_count
+
+        if is_allowed:
+            # if allowed, discount the number of blacklisted cases served to the user
+            request.user.update_case_allowance(case_count=blacklisted_count)
```

Comment:
Oh, thinking about this over the weekend, I think it doesn't work to update case allowance here. For example it doesn't take pagination into account, so we end up using up your entire case allowance even if you only request the first 10 cases. In order to take everything like that into account, you have to manually recreate the whole view logic here -- better to just hook in after it runs.